### PR TITLE
[SPARK-14718][SQL] Avoid mutating ExprCode in doGenCode

### PR DESCRIPTION
**Pull request by sameeragarwal**
## What changes were proposed in this pull request?

The `doGenCode` method currently takes in an `ExprCode`, mutates it and returns the java code to evaluate the given expression. It should instead just return a new `ExprCode` to avoid passing around mutable objects during code generation.
## How was this patch tested?

Existing Tests

### CODE REVIEW

**Comment by rxin** [comment]
cc @davies

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala [hunk 1]
```diff
@@ -104,14 +104,14 @@ case class SortPrefix(child: SortOrder) extends UnaryExpression {
...
 
-    childCode.code +
-    s"""
-      |long ${ev.value} = ${nullValue}L;
-      |boolean ${ev.isNull} = false;
-      |if (!${childCode.isNull}) {
-      |  ${ev.value} = $prefixCode;
-      |}
-    """.stripMargin
+    ev.copy(childCode.code +
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala]: put code explicitly here?

Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala]: sure, added

**Comment by rxin** [comment]
LGTM pending tests.

**Comment by rxin** [comment]
Merging in master. Thanks.

**Pull request closed without merging**